cocaine/detail/secservice.py

```diff
@@ -0,0 +1,96 @@
+#
+# Mostly compy-pasted from cocaine-tools
+#
+from tornado import gen
+from tornado.ioloop import IOLoop
+
+from cocaine.services import Service
+from cocaine.exceptions import CocaineError
+
+
+class SecureServiceError(CocaineError):
+    pass
+
+
+class BasicSecurity(object):
+    pass
+
+
+class Promiscuous(BasicSecurity):
+    @gen.coroutine
+    def fetch_token(self):
+        raise gen.Return('')
+
+
+class TVM(BasicSecurity):
+
+    def __init__(self, client_id, client_secret, name='tvm'):
+        self._client_id = client_id
+        self._client_secret = client_secret
+
+        self._tvm = Service(name)
+
+    @classmethod
+    def ty(self):
+        return 'TVM'
+
+    @gen.coroutine
+    def fetch_token(self):
+        grant_type = 'client_credentials'
+
+        channel = yield self._tvm.ticket_full(self._client_id, self._client_secret, grant_type, {})
+        ticket = yield channel.rx.get()
+
+        raise gen.Return(self._make_header(ticket))
+
+    def _make_header(self, ticket):
+        return '{} {}'.format(self.ty(), ticket)
+
+
+class SecureService(object):
+
+    def __init__(self, wrapped, secure, tok_update_sec, loop=IOLoop.current()):
+        self._wrapped = wrapped
+        self._secure = secure
+
+        self._tok_update_sec = tok_update_sec
+
+        loop.spawn_callback( lambda: self._refresh_token())
+
+    @gen.coroutine
+    def _refresh_token(self):
+        while True:
+            try:
+                self._token = yield self._secure.fetch_token()
+            finally:
+                yield gen.sleep(self._tok_update_sec)
+
+    @gen.coroutine
+    def connect(self, traceid=None):
+        yield self._wrapped.connect(traceid)
+
+    def disconnect(self):
+        return self._wrapped.disconnect()
+
+    def __getattr__(self, name):
+        @gen.coroutine
+        def wrapper(*args, **kwargs):
+            try:
+                kwargs['authorization'] = self._token
```

Comment:
I think you might have an infinite loop here if self._token has not been assigned
What exception do you suggest?